File: MATLAB/proposed_method/gtpf.m
function [x_est, choice] = gtpf(y, coeffs, fns, noise, M, B)

% Extract coeffs
C = coeffs{1};
H = coeffs{2};

% Get dimensions
T = length(y(1,:));
dy =length(y(:,1));
dx = length(C(1,:));

% Functions
g = fns{1};
h = fns{2};

% Noise
[var_x, var_y, var] = noise{:};

% Initialize
x_particles = rand(dx,M);
x_old = x_particles;
x_est = zeros(dx, T);

% Beta selection initialization
beta = 0.2;
b_size = length(B);
beta_post = ones(1,b_size)./b_size;

% Initialize weights
w = ones(1,M)/M;

% Store beta
choice = zeros(1,T);
choice(1) = beta;

% Repetitive term used in computations of weights (log form)
ln_coeff = - 0.5*dx*log(2*pi*var_y)  - 0.5*dx*log(2*pi*var_x);

% Start filter
for t=2:T

    % FIRST STAGE_________________________________________________________
    % Propose iniital particles based on model transition
    for m = 1:M
        tr_mean(:,m) = C*g(x_old(:,m));
        x_particles(:,m) = mvnrnd(tr_mean(:,m), var_x*eye(dx))';
    end

    % Predictions from proposed particles
    x_predicted = mean(x_particles,2);

    % Initialize Set of Data indices S
    states_y = 1:dy;

    % Modify proposal
    for i = 1:dx

        % If all data points have been used, reset S
        if (isempty(states_y))
            states_y = 1:dy;
        end

        % Sample at uniform a data index l
        l = datasample(states_y, 1);

        % Update S by removing sampled index
        states_y = setdiff(states_y, l);

        % For the pair (y_l, x_i(m)) find ML x_i(m*)
        for m = 1:M
            % Use the vector of predictions and replace the mth
            % particle for evaluation in the likelihood
            x_predicted(i) = x_particles(i,m);

            % Compute loglikelihood
            ln_p(m) = - (0.5/var_y)*(y(l,t) - H(l,:)*h(x_predicted) ).^2;
        end

        % Find max
        if (length(find(ln_p == max(ln_p))) ~= 1)
            % Sample at random if all weigths same
            % (Avoid numerical issues)
            m_star(i) = datasample(1:M, 1);
        else
            m_star(i) = find(ln_p == max(ln_p));
        end

        % Form proposed mean from particles with ML
        x_predicted(i) = x_particles(i, m_star(i));

    end


    % BETA posterior computation___________________________________________
    % Collection of terms needed for computation
    xy_info = {y, x_predicted, tr_mean, w};
    beta_info = {B, b_size, beta_post};
    terms = {dx, M, t, ln_coeff};

    % Call function to compute beta posterior and sample beta
    [beta, x_particles, w, beta_post] = beta_posterior(xy_info, beta_info, fns, coeffs, noise, terms);

    % Store beta sample
    choice(t) = beta;


    % SECOND STAGE_________________________________________________________
    % Propose new particles
    % We use the particles and weights stored when finding beta since
    % it's the same steps, in order to avoid double sampling
    % and evaluation

    % Resample and set weights to be equal
    idx = datasample(1:M, M, 'Weights',w);
    w = ones(1,M)/M;

    % Set new particles
    x_particles = x_particles(:,idx);
    x_old = x_particles;

    % State estimates
    x_est(:,t) = mean(x_particles,2);


end

end

File: Python/gtpf.py
import numpy as np
import random as rnd
from posterior_compute import beta_posterior

def generalized_tpf(model_info, dim, M, fns, noise,  beta_info):

    # Unpack data and model
    y, C, H = model_info

    # Unpack size variables and noise
    dx, dy, T = dim
    var_x, var_y, var_c, var_h, var = noise

    # Unpack functions
    f,g = fns

    # beta inits
    B, b_size, beta_post, chosen_beta = beta_info

    # Repetitive term
    ln_coeff = - 0.5 * dx * np.log(2 * np.pi * var_y) - 0.5 * dx * np.log(2 * np.pi * var_x)

    # Initialize
    x_particles = np.zeros((dx, M))
    tr_mean = np.zeros((dx, M))
    w = np.ones(M) / M
    x_old = x_particles
    x_est = np.zeros((dx, T))
    covX = var_x * np.eye(dx)

    for t in range(1, T):

        m_star = []
        # FIRST STAGE
        for m in range(M):
            tr_mean[:, m] = C @ f(x_old[:, m])
            x_particles[:, m] = np.random.multivariate_normal(tr_mean[:, m], covX)

        # Predictions from proposed particles
        x_predicted = np.mean(x_particles, axis=1)

        # Initialize Set of Data indices S
        states_y = np.arange(dy)

        # Modify proposal
        for i in range(dx):

            ln_p = []

            # If all data points have been used
            if states_y.size == 0:
                states_y = np.arange(dy)

            # Sample at uniform a data index l
            l = rnd.sample(list(states_y), 1)

            # Update S by removing sampled index
            states_y = np.setdiff1d(states_y, l)

            for m in range(M):
                # Use vector of predictions and replace mth particle
                x_predicted[i] = x_particles[i, m]

                # Compute loglikelihood
                ln_p.append(float(-(0.5 / var_y) * ((y[l, t] - H[l, :] @ g(x_predicted))) ** 2))

            # Find max
            if float(min(ln_p) == max(ln_p)):
                # Sample at random if all weights same (Avoid numerical issues)
                m_star.append(rnd.sample(list(range(M)), 1)[0])
            else:
                ln_p_array = np.array(ln_p)
                m_star.append(int(np.where(ln_p_array == max(ln_p_array))[0][0]))

            # Form proposed mean from particles with ML
            x_predicted[i] = x_particles[i, m_star[i]]

            # BETA posterior computation___________________________________________
            # Collection of terms needed for computation
            xy_info = [y, x_predicted, tr_mean, w]
            beta_info = [B, b_size, beta_post]
            terms = [M, t, ln_coeff]

            # Call function to compute beta posterior and sample beta
            [beta, x_particles, w, beta_post] = beta_posterior(xy_info, beta_info, dim, C, H, fns, noise, terms)

            # Store beta sample
            chosen_beta.append(beta)

            # Resample and set weights to be equal
            idx = np.random.choice(list(range(M)), M, p=w)
            w = np.ones(M) / M

            # Set new particles
            x_particles = x_particles[:, idx]
            x_old = x_particles

            # State estimates
            x_est[:, t] = np.mean(x_particles, axis=1)

    return x_est, chosen_beta